sasl/hmac.go

```diff
@@ -0,0 +1,230 @@
+package sasl
+
+import (
+	"bytes"
+	"crypto/cipher"
+	"crypto/des"
+	"crypto/hmac"
+	"crypto/md5"
+	"crypto/rc4"
+	"encoding/binary"
+	"fmt"
+)
+
+var (
+	errEmptyBuffer                = fmt.Errorf("ERROR: Empty buffer")
+	errBufferTooSmall             = fmt.Errorf("ERROR: Given buffer is too small")
+	errPeerSequenceNumberMismatch = fmt.Errorf("ERROR: Peer sequence number mismatch")
+	errInvalidPeerHMAC            = fmt.Errorf("ERROR: Invalid peer HMAC")
+)
+
+type cipherType int
+
+const (
+	cipherNone cipherType = iota
+	cipherRC4_40
+	cipherRC4_56
+	cipherRC4
+	cipherDES
+	cipher3DES
+	_cipherCount
+)
+
+type securityIntegrityContext struct {
+	SeqNum     uint32
+	PeerSeqNum uint32
+	Ki         []byte
+	PeerKi     []byte
+	EncKey     []byte
+	DecKey     []byte
+	HA1        []byte
+}
+
+type securityPrivacyContext struct {
+	securityIntegrityContext
+	cipher    cipherType
+	blockSize int
+	EncCipher func(dst, src []byte)
+	DecCipher func(dst, src []byte)
+}
+
+func (sec *securityPrivacyContext) generateKeys() error {
+	ha1len := 16
+	if sec.cipher == cipherRC4_40 {
+		ha1len = 5
+	} else if sec.cipher == cipherRC4_56 {
+		ha1len = 7
+	}
+
+	sec.EncKey = md5Sum(sec.HA1, []byte(sealingClientServer))
+	sec.DecKey = md5Sum(sec.HA1, []byte(sealingServerClient))
+	sec.Ki = md5Sum(sec.HA1[0:ha1len], []byte(signingClientServer))
+	sec.PeerKi = md5Sum(sec.HA1[0:ha1len], []byte(signingServerClient))
+
+	switch sec.cipher {
+	case cipherRC4_40, cipherRC4_56, cipherRC4:
+		sec.blockSize = 1
+		cipher, err := rc4.NewCipher(sec.DecKey)
+		if err != nil {
+			return err
+		}
+		sec.DecCipher = cipher.XORKeyStream
+		cipher, err = rc4.NewCipher(sec.EncKey)
+		if err != nil {
+			return err
+		}
+		sec.EncCipher = cipher.XORKeyStream
+	case cipher3DES:
+		block, err := des.NewTripleDESCipher(sec.DecKey)
+		if err != nil {
+			return err
+		}
+		sec.blockSize = block.BlockSize()
+		sec.DecCipher = block.Decrypt
+		block, err = des.NewTripleDESCipher(sec.EncKey)
+		if err != nil {
+			return err
+		}
+		sec.EncCipher = block.Encrypt
+	case cipherDES:
+		block, err := des.NewCipher(sec.DecKey)
+		if err != nil {
+			return err
+		}
+		sec.blockSize = block.BlockSize()
+		sec.DecCipher = cipher.NewCTR(block, sec.DecKey[len(sec.DecKey)-8:]).XORKeyStream
+		block, err = des.NewCipher(sec.EncKey)
+		if err != nil {
+			return err
+		}
+		sec.EncCipher = cipher.NewCTR(block, sec.EncKey[len(sec.EncKey)-8:]).XORKeyStream
+	default:
+		return fmt.Errorf("No such cipher")
```

Comment:
Lowercase `No`, and then show the incorrect cipher type in the error message perhaps?